### PR TITLE
Mac kext review fixes

### PR DESCRIPTION
**Pull request by pmj**
While reviewing the kext code for potential security issues in preparation for #528, I found a few places in the code which aren't ideal, although they probably aren't exploitable.

1. The memory getter & notification port setter in the `IOUserClient` subclasses were not protected by the mutex. It's probably not a serious problem, but doesn't do any harm to fix this.
2. `GetRelativePath()` can be called with 2 unrelated paths. See issue #544. The immediate upshot is that you can easily KP the system by posing as a provider, creating a new root, and renaming it. We'll need to decide what to do about renaming virtualisation roots, but I've changed `GetRelativePath()` to behave more reasonably by returning `nullptr` if the arguments don't make sense, which appears to be safe at the sole call site.
3. I noticed some logging that could be improved by using the path-based logging macro.

### CODE REVIEW

**Comment by pmj** [comment]
Some notes:

* I'm not entirely done with the review, this is just what I found specifically looking for user space API issues and string handling problems.
* `GetRelativePath()` seems like the kind of function for which we could easily create unit tests for edge cases. Before I can do this, we need to resolve the questions I've previously raised in #500.
* I still need to do some manual testing on the code for this PR.

**Comment by pmj** [comment]
I've done some testing and the `GetRelativePath()` changes do appear to successfully defend against the 3 edge cases:

 * Path shorter than root path (buffer overflow)
 * Root path not a prefix of path
 * Root path not a parent of path

So I guess this PR is now ready for review.

**Comment by wilbaker** [comment]
> GetRelativePath() seems like the kind of function for which we could easily create unit tests for edge cases.  Before I can do this, we need to resolve the questions I've previously raised about unit testing native Mac code in #500.

Some unit tests here would be great, let me catch up on the discussion in #500.  If we don't already have a work item for adding unit tests for the kext code I can add one.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/PrjFSProviderUserClient.cpp [hunk 1]
```diff
@@ -118,13 +118,20 @@ IOReturn PrjFSProviderUserClient::clientMemoryForType(
...
+            
+            Mutex_Acquire(this->dataQueueWriterMutex);
             {
-                queueMemory->retain(); // Matched internally in IOUserClient
-                *memory = queueMemory;
-                return kIOReturnSuccess;
+                queueMemory = this->dataQueueMemory;
+                if (queueMemory != nullptr)
+                {
+                    queueMemory->retain(); // Matched internally in IOUserClient
```
[the remark this answers is not in the document]
Comment (pmj) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/PrjFSProviderUserClient.cpp]: After calling `clientMemoryForType()`, a mapping to user space is made with the returned memory descriptor, after which the latter is released:

https://github.com/apple/darwin-xnu/blob/0a798f6738bc1db01281fc08ae024145e84df927/iokit/Kernel/IOUserClient.cpp#L1645

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp [hunk 2]
```diff
@@ -486,13 +486,24 @@ bool VirtualizationRoot_VnodeIsOnAllowedFilesystem(vnode_t vnode)
 
 static const char* GetRelativePath(const char* path, const char* root)
 {
-    assert(strlen(path) >= strlen(root));
+    size_t rootLength = strlen(root);
+    size_t pathLength = strlen(path);
+    if (pathLength < rootLength || 0 != memcmp(path, root, rootLength))
```
[the remark this answers is not in the document]
Comment (pmj) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: I've now tested that change thoroughly and it's working as expected. I've added one more commit that exports the canonical path as a property on the user client's registry entry, so you can see which user client is the provider for which virtualisation root in IORegistryExplorer/`ioreg`.

**Comment by pmj** [comment]
> Some unit tests here would be great, let me catch up on the discussion in #500. If we don't already have a work item for adding unit tests for the kext code I can add one.

As far as I'm concerned, this PR is ready to go once commits 4 & 5 have been reviewed, do you want to keep it open until we've got unit tests?

**Comment by wilbaker** [comment]
> As far as I'm concerned, this PR is ready to go once commits 4 & 5 have been reviewed, do you want to keep it open until we've got unit tests?

Thanks I'll take a look, I think we can add unit tests in a follow up PR that adds unit test infrastructure for the kext.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp [hunk 3]
```diff
@@ -352,50 +352,60 @@ VirtualizationRootResult VirtualizationRoot_RegisterProviderForPath(PrjFSProvide
         }
         else
         {
-            FsidInode vnodeIds = Vnode_GetFsidAndInode(virtualizationRootVNode, vfsContext);
-            uint32_t rootVid = vnode_vid(virtualizationRootVNode);
+            char virtualizationRootCanonicalPath[PrjFSMaxPath] = "";
+            int virtualizationRootCanonicalPathLength = sizeof(virtualizationRootCanonicalPath);
+            err = vn_getpath(virtualizationRootVNode, virtualizationRootCanonicalPath, &virtualizationRootCanonicalPathLength);
```
Comment (wilbaker) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: Super minor:  Update `// Return values:` comment to include that this function can return error value(s) returned by `vn_getpath`.

Also, do you know if our call to `vnode_lookup` will help prevent `vn_getpath` failures?  (I wasn't sure if there was any relationship between the name cache and `vnode_lookup`).

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: > I've just pushed a fourth commit to this PR which evaluates the root path using vn_getpath() during provider registration. I still need to do some manual testing on this change though, for which it's now a bit late here, so I'll do that tomorrow.

What is the advantage of this approach over using a case-insensitive string comparison (in place of memcpy)?  Is it a perf concern?

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp [hunk 4]
```diff
@@ -352,50 +352,60 @@ VirtualizationRootResult VirtualizationRoot_RegisterProviderForPath(PrjFSProvide
...
-                        virtualizationRootVNode = NULLVP; // prevent vnode_put later; active provider should hold vnode reference
                     
-                        KextLog_Note("VirtualizationRoot_RegisterProviderForPath: new root not found in offline roots, inserted as new root with index %d. path '%s'", rootIndex, virtualizationRootPath);
-                    }
-                    else
+                    if (0 == err)
                     {
-                        // TODO: scan the array for roots on mounts which have disappeared, or grow the array
-                        KextLog_Error("VirtualizationRoot_RegisterProviderForPath: failed to insert new root");
+                        userClient->setProperty(PrjFSProviderPathKey, virtualizationRootCanonicalPath);
```
Comment (wilbaker) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: Minor, could this code (for `setProperty`) be moved outside of holding `s_virtualizationRootsLock`?

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp [hunk 5]
```diff
@@ -352,50 +352,60 @@ VirtualizationRootResult VirtualizationRoot_RegisterProviderForPath(PrjFSProvide
...
-                        root.providerPid = clientPID;
-                        virtualizationRootVNode = NULLVP; // transfer ownership
+                        rootIndex = InsertVirtualizationRoot_Locked(userClient, clientPID, virtualizationRootVNode, rootVid, vnodeIds, virtualizationRootCanonicalPath);
+                        if (rootIndex >= 0)
+                        {
+                            assert(rootIndex < s_maxVirtualizationRoots);
+                        
+                            virtualizationRootVNode = NULLVP; // prevent vnode_put later; active provider should hold vnode reference
+                        
+                            KextLog_Note("VirtualizationRoot_RegisterProviderForPath: new root not found in offline roots, inserted as new root with index %d. path '%s'", rootIndex, virtualizationRootPath);
```
Comment (wilbaker) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: In the log message use `virtualizationRootCanonicalPath` rather than `virtualizationRootPath` (or include both if you think we'd potentially care about them being different).

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp (hunk 4 above)
Comment (wilbaker) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: What is the purpose of setting this properly on `userClient`, is it used anywhere?

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp [hunk 6]
```diff
@@ -486,13 +496,24 @@ bool VirtualizationRoot_VnodeIsOnAllowedFilesystem(vnode_t vnode)
...
+        return nullptr;
+    }
     
-    const char* relativePath = path + strlen(root);
+    const char* relativePath = path + rootLength;
     if (relativePath[0] == '/')
     {
         relativePath++;
     }
+    else if (rootLength > 0 && root[rootLength - 1] != '/' && pathLength > rootLength)
```
Comment (wilbaker) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: Can this check:

`else if (rootLength > 0 && root[rootLength - 1] != '/' && pathLength > rootLength)`

Be simplified to:

`else if (rootLength != pathLength)`

My thinking was that the root path can only be a parent of `path` at this point  if either:

- `relativePath[0] == '/'` is true
- `rootLength == pathLength` (i.e. `relativePath[0]` is `'\0'`)

Can you think of a scenario I'm missing?

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmj) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: A case insensitive string comparison is neither correct here, nor efficient.

Correctness:
 * `strcasecmp()` only takes into account ASCII characters' case, and isn't locale-specific. It's also not what case-insensitive HFS+ and APFS use.
 * Not all components of the path necessarily reside on case-insensitive file systems.
 * With all paths under consideration now being obtained via `vn_getpath()`, there's no reason there would be a case discrepancy unless a rename operation had taken place.

Efficiency:
 * Case-insensitive comparison is more computationally intensive than simple equality.
 * `memcmp()` can in theory be more efficient than the `strcmp()` family of functions as the latter must not dereference the incoming pointers past a `\0` byte, meaning the only allowed way to implement is is to work byte-by-byte. As we don't care about ordering, I suppose `bcmp()` could be even more efficient, although in practice the kernel versions of none of these functions are optimised for speed.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp (hunk 3 above)
Comment (pmj) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: It might reduce the likelihood, but as we aren't locking the name cache, nothing can be guaranteed anyway.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp (hunk 4 above)
Comment (pmj) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: It's a diagnostic aid - you can look in the IO Registry using `ioreg` or IORegistryExplorer and check the path associated with each provider. (This is explained in the commit message IIRC.) I suppose you could argue we should only set it in debug builds, but I don't see any harm, and the updates are infrequent, so not a performance concern.

Comment (pmj) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: Yes, that's safe as long as we don't start updating it elsewhere. My understanding is that we're not going to bother with the more comprehensive root rename/move support from the [refresh-root-path branch](https://github.com/pmj/VFSForGit/tree/refresh-root-path) at least for now, where the thread safety of that becomes slightly more relevant.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp (hunk 6 above)
Comment (pmj) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: In principle, if root is `"/"`, and path is `"/something/else"`, `relativePath[0]` will not be `'/'` and the lengths won't match. Although of course at the moment that would mean a virtualisation root at /, which isn't terribly practical. I don't *think* `vn_getpath()` should be trailing slashes for any other kind of path, so assuming we'll never call `GetRelativePath()` with `root = "/"` we can make that change. I'm not sure it's worth doing that for performance's sake, considering `relativePath[0] == '/'` will be the hottest code path by far.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp (hunk 5 above)
Comment (pmj) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: Fixed, thanks!

File: ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp (hunk 6 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: > In principle, if root is "/", and path is "/something/else", relativePath[0] will not be '/' and the lengths won't match.

Ahh, I'd been assuming that `root` would never end with `"/"` (even in the case of the root directory).

Comment (pmj) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: I've definitely seen `vn_getpath` return `"/"` for the root vnode. The empty path would be relative, not absolute, so I guess that's why.

Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp]: > I've definitely seen vn_getpath return "/" for the root vnode. The empty path would be relative, not absolute, so I guess that's why.

Ok cool, thanks again for the details!
